File: src/2020/python/src/day06.py
def star1():
    result = 0
    yes = set()

    for line in open("../inputs/day6.txt", "r"):
        line = line.strip()

        if line == "":
            result += len(yes)
            yes.clear()
        else:
            for i in line:
                yes.add(i)

    return result


def star2():
    result = 0
    responses = []

    for line in open("../inputs/day6.txt", "r"):
        line = line.strip()

        if line == "":
            if responses:
                temp = responses[0]
                for r in responses[1:]:
                    temp = temp.intersection(r)
                result += len(temp)
                responses.clear()
        else:
            temp = set()
            for i in line:
                temp.add(i)
            responses.append(temp)

    return result


print("STAR 1: {}\nSTAR 2: {}".format(star1(), star2()))

File: src/2020/rust/src/day_06.rs
use std::collections::HashSet;

use crate::utils;

fn star1() -> usize {
    let mut result: usize = 0;
    let mut yes: HashSet<char> = HashSet::new();
    let lines = utils::file_to_lines("../inputs/day6.txt");

    for line in lines {
        if line == "" {
            result += yes.len();
            yes.clear();
        } else {
            for i in line.chars() {
                yes.insert(i);
            }
        }
    }

    return result;
}

fn star2() -> usize {
    let mut result: usize = 0;
    let mut responses: Vec<HashSet<char>> = Vec::new();
    let lines: Vec<String> = utils::file_to_lines("../inputs/day6.txt");

    for line in lines {
        if line == "" {
            if !responses.is_empty() {
                let mut temp = responses[0].clone();
                for r in responses[1..].iter() {
                    temp = temp.intersection(r).copied().collect();
                }
                result += temp.len();
                responses.clear();
            }
        } else {
            let mut temp: HashSet<char> = HashSet::new();
            for i in line.chars() {
                temp.insert(i);
            }
            responses.push(temp);
        }
    }

    return result;
}

pub fn main() {
    println!("DAY 6\n=====\nSTAR 1: {}\nSTAR 2: {}\n", star1(), star2());
}